EpikCord/client/websocket_client.py

```diff
@@ -48,198 +58,231 @@ def __init__(
+                wait_for_callback[0].set_result(data)
+                self.wait_for_events[event_name].remove(wait_for_callback)
+
+    async def dispatch(self, event_name: str, *args, **kwargs):
+        for callback in self.events[event_name]:
+            await callback(*args, **kwargs)
+
+    def wait_for(
         self,
-        guild_id: int,
+        event_name: str,
         *,
-        query: Optional[str] = None,
-        limit: Optional[int] = None,
-        presences: Optional[bool] = None,
-        user_ids: Optional[List[str]] = None,
-        nonce: Optional[str] = None,
+        check: Optional[Callback] = None,
+        timeout: Union[float, int] = 0,
     ):
-        payload: dict = {
-            "op": GatewayOpcode.REQUEST_GUILD_MEMBERS,
-            "d": {"guild_id": guild_id},
-        }
+        """
+        Waits for the event to be triggered.
 
-        if query:
-            payload["d"]["query"] = query
+        Parameters
+        ----------
+        event_name : str
+            The name of the event to wait for.
+        check : Optional[callable]
+            A check to run on the event.
+            If it returns ``False``, the event will be ignored.
+        timeout : int
+            The amount of time to wait for the event.
+            If not specified, it'll wait forever.
+        """
+        future: asyncio.Future = asyncio.Future()
 
-        if limit:
-            payload["d"]["limit"] = limit
+        if not check:
 
-        if presences:
-            payload["d"]["presences"] = presences
+            async def check(*_, **__): # type: ignore
+                return True
 
-        if user_ids:
-            payload["d"]["user_ids"] = user_ids
+        self.wait_for_events[event_name.lower()].append((future, check))
+        return asyncio.wait_for(future, timeout=timeout)
 
-        if nonce:
-            payload["d"]["nonce"] = nonce
+    def event(self, event_name: Optional[str] = None):
+        def register_event(func):
+            func_name = event_name or func.__name__.lower()
 
-        await self.send_json(payload)
+            if func_name.startswith("on_"):
+                func_name = func_name[3:]
 
-    async def reconnect(self):
-        await self.close()
-        await self.connect()
-        await self.identify()
-        await self.resume()
+            self.events[func_name].append(func)
+
+            return Event(func, event_name=func_name)
+
+        return register_event
 
     async def handle_close(self):
-        if self.ws.close_code == GatewayCECode.DisallowedIntents:
+        if self.websocket.close_code == GatewayCECode.DisallowedIntents:
             raise DisallowedIntents(
                 "You cannot use privileged intents with this token, go to "
                 "the developer portal and allow the privileged intents "
                 "needed. "
             )
-        elif self.ws.close_code == 1006:
+        elif self.websocket.close_code == 1006:
```

Comment:
`1006` should be extracted to a constant or a enum entry

Reply:
Maybe a `WebsocketCloseCode`?